### ISSUE TITLE
Create encoding 

### ISSUE THREAD

**Issue by sharonzagh** [jpmml/sklearn2pmml#252]
Hi, 
I try to convert a feature to lower case and then create an encoding of this new feature but it appears that it creates the encoding based on the original data 
    
```
def categorical_str_transform(mapper: DataFrameMapper, features: list,replace='999'):
    feature_defs = [
    (col,[CategoricalDomain(missing_value_treatment="as_value",
                                   missing_value_replacement=replace,
                                   with_statistics=False,
                                   with_data=False,
                                  ),
          StringNormalizer(function = "lowercase")]
    ) 
        for col in features
    ]
    mapper.features = mapper.features + feature_defs

def add_expression2(mapper, input_fields: list, expression: str, alias: str):
    feature_defs =  [
        (col , [ExpressionTransformer(expression), 
                Alias(CastTransformer(float), name=alias)
               ],
        {'alias': alias}
    )
        for col in input_fields
    ]
    mapper.features = mapper.features + feature_defs

df_train = pd.DataFrame()

df_train['feature1'] = [1, 1, 'Released', 1, 1, 1, 1, 'Released', 1, 1] 

recorder2= DataFrameMapper([])
recorder2.features = []

categorical_str_transform(recorder2,['feature1'])

str_expression_test = " 1 if X[0] == 'released' else 0"

add_expression2(recorder2, [['feature1']],str_expression_test, 'feature1_released')

recorder2.df_out = True
res = recorder2.fit_transform(df_train ,y_train)
```

the res  df will have the values of feature1 in lower case but feature1_released will have only zeros because it will compare to the original dataset that doesn't have the value 'released' but 'Released' instead 

Do you know how we can handle it in order to have the encoding based on the dataset in lower case

**Comment by vruusmann** [comment]
> Do you know how we can handle it in order to have the encoding based on the dataset in lower case

Apply `StringNormalizer(function = "lowercase")` **right before** the encoding step?

What's the issue here? Are you claiming that the `StringNormalizer` does not convert string values to lowercase, or are you simply asking me to refactor some Python application code? Clearly, this issue does not have any relation to the (J)PMML portion of the workflow.

**Comment by vruusmann** [comment]
If I do `print(res)` then I see that `"released"` is printed in lowercase letters.

**Comment by vruusmann** [comment]
The syntax `X[0]` is performing string slicing in Python's `eval()` function:

```python
df_train = pd.DataFrame()
df_train['feature1'] = [1, "Released"] 

mapper = DataFrameMapper([
	("feature1", [CastTransformer(dtype = str), StringNormalizer(function = "lowercase"), ExpressionTransformer("X[0]")])
])
res2 = mapper.fit_transform(df_train)
print(res2)
```

Adjust your solution accordingly.

**Comment by vruusmann** [comment]
> The syntax X[0] is performing string slicing in Python's eval() function

Or maybe I'm wrong, and Pandas' `DataFrame.append()` function is simply performing string truncation here?

The fact of the matter is that `ExpressionTransformer` is seeing a 1-character strings there.

**Comment by vruusmann** [comment]
> Or maybe I'm wrong, and Pandas' DataFrame.append() function is simply performing string truncation here?

No, it's still an unintentional string slicing issue, because switching to `ExpressionTransformer("X[:]")` prints full-length strings.

**Comment by vruusmann** [comment]
TLDR: The `X[<index>]` syntax works differently for string vs numeric columns.

Related to https://github.com/jpmml/jpmml-python/issues/6
